### PR TITLE
os/bluestore: fix bmap allocator relevant issues

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas Minor fixes about the newly Bmap allocator. Please help to review, thanks:-)

**Comment by liewegas** [comment]
@chhabaramesh can you review this?

**Comment by xiexingguo** [comment]
retest this please

**Comment by chhabaramesh** [comment]
@liewegas  @xiexingguo  I will review it today.

File: src/os/bluestore/BitMapAllocator.cc [hunk 1]
```diff
@@ -172,7 +172,7 @@ void BitMapAllocator::init_rm_free(uint64_t offset, uint64_t length)
     " offset " << offset << " length " << length << dendl;
 
   assert(!(offset % m_block_size));
-  assert(!(offset % m_block_size));
```
Comment (chhabaramesh) [inline src/os/bluestore/BitMapAllocator.cc]: looks good.

File: src/os/bluestore/BitMapAllocator.cc [hunk 2]
```diff
@@ -88,8 +88,7 @@ int BitMapAllocator::allocate(
 {
 
   assert(!(alloc_unit % m_block_size));
-  int64_t len = NEXT_MULTIPLE(want_size, m_block_size);
-  int64_t nblks = len / m_block_size;
+  int64_t nblks = (want_size + m_block_size - 1) / m_block_size;
```
Comment (chhabaramesh) [inline src/os/bluestore/BitMapAllocator.cc]: looks good

File: src/os/bluestore/BitMapAllocator.cc [hunk 3]
```diff
@@ -31,9 +31,6 @@ BitMapAllocator::BitMapAllocator(int64_t device_size)
   m_bit_alloc = new BitAllocator(device_size / block_size,
         zone_size_blks, CONCURRENT, true);
   assert(m_bit_alloc);
-  if (!m_bit_alloc) {
-    dout(10) << __func__ << "Unable to intialize Bit Allocator" << dendl;
```
Comment (chhabaramesh) [inline src/os/bluestore/BitMapAllocator.cc]: I would suggest to change order. Also assert can be disabled non debug build. So it is not truly unreachable.

File: src/os/bluestore/BitAllocator.cc [hunk 4]
```diff
@@ -935,7 +935,8 @@ bool BitAllocator::is_allocated(int64_t start_block, int64_t num_blocks)
 bool BitAllocator::is_allocated(int64_t *alloc_blocks, int64_t num_blocks)
 {
   for (int64_t i = 0; i < num_blocks; i++) {
-    return is_allocated(alloc_blocks[i], 1);
+    if (!is_allocated(alloc_blocks[i], 1))
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

File: src/os/bluestore/BitAllocator.cc [hunk 5]
```diff
@@ -339,7 +339,7 @@ int64_t BmapList::incr_marker(int64_t add)
 
 void BmapList::set_marker(int64_t val)
 {
- atomic_store(&m_marker, val);
+  std::atomic_store(&m_marker, val);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

File: src/os/bluestore/BitAllocator.cc [hunk 6]
```diff
@@ -128,7 +128,7 @@ bmap_t BmapEntry::bit_mask(int bit)
 }
 bool BmapEntry::check_bit(int bit)
 {
-  return (m_bits & bit_mask(bit));
+  return (atomic_fetch() & bit_mask(bit));
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

File: src/os/bluestore/BitAllocator.cc [hunk 7]
```diff
@@ -163,10 +163,12 @@ void BmapEntry::clear_bits(int offset, int num_bits)
 
 void BmapEntry::set_bits(int offset, int num_bits)
 {
-  for (int i = 0; i < num_bits; i++) {
-    bmap_t bmask = bit_mask(i + offset);
-    (void) std::atomic_fetch_or(&m_bits, bmask);
+  if (num_bits == 0) {
+    return;
   }
+
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

**Comment by chhabaramesh** [comment]
@liewegas  @xiexingguo 
Except change 
https://github.com/ceph/ceph/pull/9162/commits/026f96e8b86f3b82804d965374b253930d7e0369. 
Rest looks ok. 

@xiexingguo  Thanks for reviewing in details and finding these.

**Comment by xiexingguo** [comment]
@chhabaramesh My pleasure and dropped that change.

I am glad that you are back:-)

**Comment by chhabaramesh** [comment]
Somehow I did not notice the notification and took so much time to review it.

Sorry for delay @liewegas  and @xiaoxichen .

**Comment by xiaoxichen** [comment]
Did this change pass QA run?

**Comment by xiaoxichen** [comment]
did this passing rados suite?

**Comment by xiexingguo** [comment]
@xiaoxichen 
There is no such test suit and is not ready for QA yet.

**Comment by yuyuyu101** [comment]
@xiexingguo 
except no one is responsible for the pr, it would be better to let maintainer to ack then merge.

**Pull request merged**